### PR TITLE
Documentation, Comments and renaming some variables and functions (fixes #385)

### LINKED ISSUE #385: Examples Documentation Enhancements
## Is your feature request related to a problem? Please describe

Adding content to the `Examples.md` documentation.

## Is this a bug? Please provide steps to reproduce, a failing test etc.

N/A

## Describe the solution

N/A

## Additional context

The `examples` directory is a great idea and is very valuable. It helped me to get the spirit of the project very fast. I tried to give my contribution to make it even more friendly for newcomers.

### PR DESCRIPTION
**Pull request by rbroggi**
Signed-off-by: Rodrigo Broggi <ro_broggi@hotmail.com>

## Which problem is this PR solving?

This PR aims on enriching the whole experience of using the (amazing) `examples` folder. During it's development some small bugs were identified and therefore I tried to solve them as well within the scope of the PR. 

You can find more detailed comments across the changed files. 

1. Documentation enhancement with creation of diagrams and new notes in the `Examples.md` file [Issue 385](https://github.com/beatlabs/patron/issues/385).
2. Creation of a wrapper script called `services.sh` in the `examples` folder that helps with startup/stop of the services in the `examples` folder
3. Attempt to correct two bugs that were affecting the correct functioning of the `examples` integration suite:
    1. panic in `amqp/main.go` ( bug in the actual prd code. [issue 387](https://github.com/beatlabs/patron/issues/387))
    2. error in decoding in `sqs/main.go#117` 

## Short description of the changes

1. Documentation enhancements
2. Minor `nil` pointer defensive programming 
3. update of the `examples` integration suite that was not working

### CODE REVIEW

**Comment by rbroggi** [comment]
I'm wondering how it's possible that the coverage decreased in a package where I did not make any changes :thinking:
Any known issue?

**Comment by tpaschalis** [comment]
Hello @rbroggi, thanks for taking the time to contribute to Patron and improve the documentation! Don't worry about the coveralls report, it's a known false flag and we should have a fix coming up. 

I'll take some time later today to properly review the rest of this PR. I really liked some of your suggestions, especially the `example-arch.plantuml` idea, I'm definitely using it for my own projects as well 😅 

Overall, I'd be interested in hearing more about your experience with Patron! Like how you found out about the project, what were your first thoughts when using it, any pain points, or novel use-cases..

**Comment by mantzas** [comment]
@tpaschalis Just to help you with the why. @rbroggi will be joining Matching soon!
@rbroggi Hey, hope that everything is ok? Will take a look at your PR over the weekend.

**Comment by rbroggi** [comment]
> @tpaschalis Just to help you with the why. @rbroggi will be joining Matching soon!
> @rbroggi Hey, hope that everything is ok? Will take a look at your PR over the weekend.

Looking forward to join you guys. Very inspiring first read of the project. Was easy to understand and the `examples` folder is very good. Just wanted to let my small contribution there.

**Comment by rbroggi** [comment]
> Hello @rbroggi, thanks for taking the time to contribute to Patron and improve the documentation! Don't worry about the coveralls report, it's a known false flag and we should have a fix coming up.
> 
> I'll take some time later today to properly review the rest of this PR. I really liked some of your suggestions, especially the `example-arch.plantuml` idea, I'm definitely using it for my own projects as well 
> 
> Overall, I'd be interested in hearing more about your experience with Patron! Like how you found out about the project, what were your first thoughts when using it, any pain points, or novel use-cases..

I'm very happy that you appreciated the uml. Here I have created the diagram source and exported it to a `png`. There is also a fashion to render the plantuml in the markdown directly but I did not use it because it is not really built in in markdown and it would not be visible in the PR itself: https://github.com/jonashackt/plantuml-markdown

File: .gitignore
Comment (rbroggi) [inline .gitignore:29]: This new entry on the `.gitignore` is to protect against committing temporary files created by the new script `examples/services.sh`.

File: client/sns/v2/publisher.go
Comment (rbroggi) [inline client/sns/v2/publisher.go:71]: As described in [issue 387](https://github.com/beatlabs/patron/issues/387) this protection corrects the behavior of the `examples/amqp/main.go`. It also makes the service more robust against panicing

File: component/http/route.go
Comment (rbroggi) [inline component/http/route.go:209]: Added the directory path as part of the error message for faster debugging.

File: component/http/route.go
Comment (rbroggi) [inline component/http/route.go:220]: Added the file path as part of the error message for faster debugging.

File: docs/Examples.md
Comment (rbroggi) [inline docs/Examples.md:33]: This simple diagram is meant to give a visual input on how the whole integration suite communicates.

File: docs/Examples.md
Comment (rbroggi) [inline docs/Examples.md:77]: I found useful making this comment because by simply trying to start both services they will fail as they attempt to listen the same port

File: docs/images/example-arch.plantuml
Comment (rbroggi) [inline docs/images/example-arch.plantuml:1]: You can use the following rendering online resource to edit and discover this file:

https://www.planttext.com/

For insights on syntax:  https://plantuml.com/component-diagram

File: examples/http-sec/main.go
Comment (rbroggi) [inline examples/http-sec/main.go:55]: renaming for more semantic meaning... 

I think it was httpCmp before as it was probably copied from the `http/main.go` file as a template.

File: examples/http/main.go
Comment (rbroggi) [inline examples/http/main.go:30]: I have introduced this environment variable so that the service http/main.go could be run from any folder and not just from the project root directory.

File: examples/services.sh
Comment (rbroggi) [inline examples/services.sh:1]: I used this script to help me debug the issue in the `examples` integration suite. 
This script does 3 main things:

1. Compile the `xxx/main.go` binaries into binaries hosted in a temporary folder. The added value was that killing the servers became easier for detached processes (let me know if this is unclear I can explain better)
2. Redirect the logs of every service started to a temporary log folder.
3. Keep track of the pids of the services processes so it becomes easy to stop/restart them.

File: examples/sqs/main.go
Comment (rbroggi) [inline examples/sqs/main.go:117]: This decoding was failing because it expects a pointer and not a value. This issue was breaking the chain of events.

File: docs/images/example-arch.plantuml
Comment (mantzas) [reply docs/images/example-arch.plantuml:1]: Nice way to create diagrams with human readable dsl.

Comment (rbroggi) [reply docs/images/example-arch.plantuml:1]: Glad that you liked it.

Most the IDEs have good plugins for live-rendering while editing the source code. Even though that can be done in the browser, I like to stay in the flow in the IDE. Here below screenshot of the plugin I use in Goland:

image

If you are interested here is the [link for the plugin home page](https://plugins.jetbrains.com/plugin/7017-plantuml-integration)

File: examples/services.sh
Comment (mantzas) [reply examples/services.sh:1]: Hey, would it not make more sense to run them with simple `go run main.go` command to keep lean and not need to create a bash script?
The idea is to open each of them in a separate terminal with `go run` and look at the logs in the terminal.

**Review by mantzas — COMMENTED** [review]
Nice contribution. I added some notes.

File: examples/services.sh
Comment (rbroggi) [reply examples/services.sh:1]: I found a bit laborious to have 7 concurrent terminal windows opened and switch between and I only created the script because I found easier to tear down / bring up the instances as well as to refer to the right logs without having to 'search' through the 7 opened terminal windows. 
I also faced issues while trying to kill detached processes started with `go run` in background mode (e.g. `go run examples/http/main.go &`) - in those cases if we kill the `go run` process it will not actually kill the server process and therefore the port allocated to the process. The same issue does not happen if the process is start straight from the compiled binary.

I understand that indeed there is the price of having another script therefore I can exclude this from the PR if you think it's not worth it.

File: docs/Examples.md
Comment (tpaschalis) [inline docs/Examples.md]: How about *"To run these examples on your system, you need the following dependencies installed:"*. 

Just avoiding the duplicate 'on your system'.

File: docs/Examples.md
Comment (tpaschalis) [inline docs/Examples.md]: s/an HTTP exposing/an HTTP component exposing/
s/readyness/readiness/

File: examples/services.sh
Comment (mantzas) [reply examples/services.sh:1]: It makes sense my only concern was the maintenance.

**Review by mantzas — DISMISSED** [review]
LGTM

**Comment by rbroggi** [comment]
@mantzas : last commit was to correct the typos spot by @tpaschalis

**Review by tpaschalis — DISMISSED** [review]
LGTM, nice work Rodrigo!

**Comment by rbroggi** [comment]
> LGTM, nice work Rodrigo!

Looking forward to reach you guys

**Review by mantzas — DISMISSED** [review]

**Comment by rbroggi** [comment]
I will have to investigate why the CI is failing. I have problems in my local environment after last update of Docker because my PC has ip6 disabled and docker commands are failing.

**Comment by rbroggi** [comment]
Dears, @mantzas  and @tpaschalis , I pushed a tiny correction to align the unit_test error message expectation to the new error message format. 

Would be great to have your review/approval again once possible. 

Thank you,
Rodrigo

**Review by mantzas — APPROVED** [review]

**Review by tpaschalis — APPROVED** [review]
LGTM, nice addition! Check out https://beatlabs.github.io/patron/ once this is merged to see your changes in action! 😄

**Pull request merged**
